### PR TITLE
Tool static analysis via compiler

### PR DESCRIPTION
**Pull request by frank-at-adacore**

### CODE REVIEW

File: courses/static_analysis_via_compiler/010_static_analysis_via_compiler.rst [hunk 1]
```diff
@@ -0,0 +1,1479 @@
...
+
+    A. 5, 6, 8, 9, 11
+    B. 11
+    C. :answer:`5, 6, 11`
+    D. No violations
+
+    .. container:: animate
+
+       + Line 5 - Dispatch needed to determine size of O
+       + Line 6 - Just a copy
```
Comment (leogermond) [inline courses/static_analysis_via_compiler/010_static_analysis_via_compiler.rst:982]: I believe this is a bit weirdly phrased, getting the value from the concrete object would need solving T's concrete type, so somewhat "dispatching". The "just a copy" does not make it seem like there is dispatching involved

**Review by leogermond — CHANGES_REQUESTED** [review]

File: courses/static_analysis_via_compiler/010_static_analysis_via_compiler.rst [hunk 2]
```diff
@@ -0,0 +1,1479 @@
...
+  + Likely generates many warnings you'll end up ignoring
+  + But you might want some of them, individually
+
+---------------------------------------------------
+Highly Optional Warnings Enabled By -gnatw.e
+---------------------------------------------------
+
++ Implicit dereferencing (missing optional :ada:`.all`)
++ Nested declaration hiding an outer declaration
++ Tags (strings) added to warning messages
```
[the remark this answers is not in the document]
Comment (frank-at-adacore) [reply courses/static_analysis_via_compiler/010_static_analysis_via_compiler.rst:322]: If you mean line 322, it's an option to add things like "warning", "error", and "info" to error message (-gnatw.d)

File: courses/static_analysis_via_compiler/010_static_analysis_via_compiler.rst [hunk 3]
```diff
@@ -0,0 +1,1479 @@
...
+
+--------------------------------
+Probable Errors - Explanations
+--------------------------------
+
+* Line 4 - Coder probably meant :ada:`2 ** 32`
+
+   * But maybe not? It could be a bit location
+
+* Line 11 - :ada:`E` is :ada:`natural`, so it should never be less than zero
```
[the remark this answers is not in the document]
Comment (frank-at-adacore) [reply courses/static_analysis_via_compiler/010_static_analysis_via_compiler.rst:254]: Changed to "can" and added something about invalid data (as in the warning message)

File: courses/static_analysis_via_compiler/010_static_analysis_via_compiler.rst [hunk 4]
```diff
@@ -0,0 +1,1479 @@
...
+
+---------------------------------------------------
+Highly Optional Warnings Enabled By -gnatw.e
+---------------------------------------------------
+
++ Implicit dereferencing (missing optional :ada:`.all`)
++ Nested declaration hiding an outer declaration
++ Tags (strings) added to warning messages
++ Record representation clauses that specify gaps
++ Overlapping subprogram actual parameters
```
[the remark this answers is not in the document]
Comment (frank-at-adacore) [reply courses/static_analysis_via_compiler/010_static_analysis_via_compiler.rst:324]: Updated list to show all added switches (as displayed via GUI)

File: courses/static_analysis_via_compiler/010_static_analysis_via_compiler.rst [hunk 5]
```diff
@@ -0,0 +1,1479 @@
...
+``TOOL_NAME ::= SPARK | GNATprove``
+
+  * Control which tool responds to pragma
+
+``REASON ::= Reason => STRING_LITERAL {& STRING_LITERAL}``
+
+  * Informational message that can be parsed by external tools
+
+--------------------------------
+Pragma Warnings Usage Examples
```
[the remark this answers is not in the document]
Comment (frank-at-adacore) [reply courses/static_analysis_via_compiler/010_static_analysis_via_compiler.rst:437]: Wildcard used in loop range example

File: courses/static_analysis_via_compiler/010_static_analysis_via_compiler.rst [hunk 6]
```diff
@@ -0,0 +1,1479 @@
...
++ Uses language-defined :ada:`pragma Restrictions`
+
+   .. code:: Ada
+
+      pragma Restrictions (restriction{, restriction});
+      restriction ::= restriction_identifier |
+                      restriction_parameter_identifier =>
+                            restriction_parameter_argument
+
++ Provides control over many features
```
[the remark this answers is not in the document]
Comment (frank-at-adacore) [reply courses/static_analysis_via_compiler/010_static_analysis_via_compiler.rst:821]: Added a line about using "pragma profile" to specify runtime and how that would implement certain restrictions

File: courses/static_analysis_via_compiler/010_static_analysis_via_compiler.rst [hunk 7]
```diff
@@ -0,0 +1,1479 @@
...
+
+----------------------------------
+Applying Restriction Identifiers
+----------------------------------
+
++ In source or in configuration file
+
+  + Configuration file name can be default or given by gpr file
+
+     + Default config file name is :filename:`gnat.adc`
```
[the remark this answers is not in the document]
Comment (frank-at-adacore) [reply courses/static_analysis_via_compiler/010_static_analysis_via_compiler.rst:893]: Yes and no. Default name works if you are NOT USING a GPR file (I just did some experimenting). So I will mention the difference

File: courses/static_analysis_via_compiler/010_static_analysis_via_compiler.rst (hunk 1 above)
Comment (frank-at-adacore) [reply courses/static_analysis_via_compiler/010_static_analysis_via_compiler.rst:982]: Cleaned up the language

**Review by leogermond — CHANGES_REQUESTED** [review]
Missing integration to CI, otherwise alright

File: courses/static_analysis_via_compiler/010_static_analysis_via_compiler.rst [hunk 8]
```diff
@@ -0,0 +1,1479 @@
+*****************************
+Static Analyis Via Compiler
```
[the remark this answers is not in the document]
Comment (frank-at-adacore) [reply courses/static_analysis_via_compiler/010_static_analysis_via_compiler.rst:2]: Done

**Review by leogermond — APPROVED** [review]

**Pull request merged**